client.go

```diff
@@ -575,146 +571,154 @@ func NewClient[TTx any](driver riverdriver.Driver[TTx], config *Config) (*Client
 			return err
 		}
-	}
 
-	go func() {
-		<-fetchCtx.Done()
-		c.signalStopComplete(ctx)
-	}()
+		if c.completer != nil {
+			// The completer is part of the services list below, but although it can
+			// stop gracefully along with all the other services, it needs to be
+			// started with a context that's _not_ fetchCtx. This ensures that even
+			// when fetch is cancelled on shutdown, the completer is still given a
+			// separate opportunity to start stopping only after the producers have
+			// finished up and returned.
+			if err := c.completer.Start(ctx); err != nil {
+				stopServicesOnError()
+				return err
+			}
 
-	c.baseService.Logger.InfoContext(workCtx, "River client successfully started", slog.String("client_id", c.ID()))
+			// Receives job complete notifications from the completer and
+			// distributes them to any subscriptions.
+			c.completer.Subscribe(c.distributeJobCompleterCallback)
+		}
 
-	return nil
-}
+		// We use separate contexts for fetching and working to allow for a graceful
+		// stop. Both inherit from the provided context, so if it's cancelled, a
+		// more aggressive stop will be initiated.
+		workCtx, c.workCancel = context.WithCancelCause(withClient[TTx](ctx, c))
 
-// ctx is used only for logging, not for lifecycle.
-func (c *Client[TTx]) signalStopComplete(ctx context.Context) {
-	for _, producer := range c.producersByQueueName {
-		producer.Stop()
-	}
+		for _, service := range c.services {
+			if err := service.Start(fetchCtx); err != nil {
+				stopServicesOnError()
+				return err
+			}
+		}
 
-	// Stop all mainline services where stop order isn't important.
-	startstop.StopAllParallel(append(
-		// This list of services contains the completer, which should always
-		// stop after the producers so that any remaining work that was enqueued
-		// will have a chance to have its state completed as it finishes.
-		//
-		// TODO: there's a risk here that the completer is stuck on a job that
-		// won't complete. We probably need a timeout or way to move on in those
-		// cases.
-		c.services,
+		for _, producer := range c.producersByQueueName {
+			producer := producer
 
-		// Will only be started if this client was leader, but can tolerate a stop
-		// without having been started.
-		c.queueMaintainer,
-	))
+			if err := producer.StartWorkContext(fetchCtx, workCtx); err != nil {
+				stopProducers()
+				stopServicesOnError()
+				return err
+			}
+		}
 
-	c.baseService.Logger.InfoContext(ctx, c.baseService.Name+": All services stopped")
+		return nil
+	}(); err != nil {
+		defer close(stopped)
+		if errors.Is(context.Cause(fetchCtx), startstop.ErrStop) {
+			return rivercommon.ErrShutdown
+		}
+		return err
+	}
 
-	// As of now, the Adapter doesn't have any async behavior, so we don't need
-	// to wait for it to stop.  Once all executors and completers are done, we
-	// know that nothing else is happening that's from us.
+	go func() {
+		defer close(stopped)
```

Comment:
It might be more readable and easily debuggable if some of these closures were split into internal client methods. Thinking in particular of the potential for a trace to have anonymous goroutines rather than named methods.

Not a huge concern but it might be worth trying to shuffle some of this around to improve the readability of `Start()`? Particularly here and in the large startup closure above.

Reply:
Took a look at this, and not against it, but I think I'm going to punt for the time being. I kind of like having all the logic `Start` (and the majority of the stop) logic in one place because you can scan it all at once without having to jump around. Also lets us reduce the number of struct variables and helpers in favor of definitions that can't leak into other contexts, and are defined more visibly, closer to home.